Update to curv@0.9

tmpfs opened the pull request.

tmpfs commented:
Also, is it possible to publish a version of `rust-paillier@0.4.2` with a view to getting rid of the `git` dependencies?

survived commented:
@tmpfs gg18 tests were failing because of [this](https://github.com/ZenGo-X/multi-party-ecdsa/pull/144/commits/e41685e68c6648167521e939c9dcd43692090855) 🙂

tmpfs commented:
> @tmpfs gg18 tests were failing because of [this](https://github.com/ZenGo-X/multi-party-ecdsa/pull/144/commits/e41685e68c6648167521e939c9dcd43692090855) slightly_smiling_face

Great catch @survived, what was I thinking adding those parentheses :disappointed:

tmpfs commented:
I will have a go at resolving the conflicts now!

tmpfs commented:
@survived, I had completely removed `zeroize` as it was redundant but after rebasing against master it was introduced again, I have removed it again under the assumption that `BigInt` also calls `zero()` on `drop` - correct me if I am wrong here please!

tmpfs commented:
@survived, rebased and fixed the compiler errors so I think that if my assumption about removing the `zeroize` attributes from the new `src/utilities/mta/range_proof.rs` file are correct then this is ready for review :+1:

tmpfs commented:
Oh, i still need to make `Keys` and `LocalKeys` generic over `Curve` too but now is probably a good time to review everything so far!

tmpfs commented:
@survived, I have now updated `Keys` and `LocalKey` to be generic over `Curve` so I think this is ready for review/merge when you have time, thanks :+1:

DmytroTym commented:
> @survived, I had completely removed `zeroize` as it was redundant but after rebasing against master it was introduced again, I have removed it again under the assumption that `BigInt` also calls `zero()` on `drop` - correct me if I am wrong here please!

I don't think this is the case tbh. From https://github.com/ZenGo-X/curv/blob/master/src/arithmetic/big_gmp.rs I get the impression that while zeroize is implemented for BigInt, it is not called on drop.

DmytroTym commented:
> I don't think this is the case tbh. From https://github.com/ZenGo-X/curv/blob/master/src/arithmetic/big_gmp.rs I get the impression that while zeroize is implemented for BigInt, it is not called on drop.

That's incorrect, if you're using `gmp` backend, `BigInt`s are actually zeroized on drop. `num-bigint` backend however does not support zeroization on drop atm, but hopefully will very soon.

tmpfs commented:
Thanks @DmytroTym for the information. I will restore the `Zeroize` attributes and dependency and raise a separate issue to remove them once `num-bigint` supports zeroization on drop.

tmpfs commented:
@survived, I finished fixing all the clippy warnings. I wasn't sure the preferred approach for using the unit type as `Error` but figured that using custom errors is better than ignoring those clippy warnings so I hope that is ok :pray:

DmytroTym commented:
Hello!
Does the code from this PR compile for everyone? One error that I get is from `gg18_sign_client`. [Here](https://github.com/tmpfs/multi-party-ecdsa/blob/curv-0.8/examples/gg18_sign_client.rs#L105) a `usize` is passed into `SignKeys::create` method, that actually [expects](https://github.com/tmpfs/multi-party-ecdsa/blob/curv-0.8/src/protocols/multi_party_ecdsa/gg_2018/party_i.rs#L389) a `u16`.

survived commented:
@DmytroTym, sound like it's me who introduced this error. I'll fix the examples, thanks!

survived commented:
@DmytroTym, fixed!

DmytroTym commented:
Hey @tmpfs and @survived. I fixed one line that prevented benchmarks from running, but otherwise imo this PR is pretty safe to merge. After all, it does not change any logic and I think that compiler and tests should have done a good job of preventing any accidental changes that would break something. Plus two pair of eyes looked at the code.
@tmpfs, as to this issue: https://github.com/ZenGo-X/multi-party-ecdsa/issues/148. As far as I understand, any efforts of zeroizing things would have to be in a new major version of curv as this PR: https://github.com/ZenGo-X/curv/pull/154 introduces breaking changes? Anyway, properly zeroizing `num-bigint` values would be tricky and probably won't be properly implemented soon. @survived please correct me if I'm wrong here. I'm not sure what should we do here, probably just give up on zeroizing `num-bigint`?

tmpfs commented:
Thanks @DmytroTym - would be great to land this :+1: Maybe just go ahead and close #148 if we are unlikely to add zeroizing in `num-bigint`.

DmytroTym commented:
@tmpfs I cannot really close the issue, we should wait for @survived :)
Meanwhile, could you please merge in https://github.com/tmpfs/multi-party-ecdsa/pull/1 so that benchmarks can run?

Hrezaei commented on benches/multi_party_ecdsa/gg18/keygen.rs:
```diff
@@ -65,11 +72,11 @@ mod bench {
+                        vec_j[i].clone()
                     })
-                    .collect::<Vec<FE>>()
+                    .collect::<Vec<Scalar<Secp256k1>>>()
```
I suggest defining `pub type FE = Scalar<Secp256k1>;` at the top of this file. In this way, when we are about to use the same code for Ed25519, we will need to change just one place instead of many. The same is requested for `pub type GE = Point<Secp256k1>;` please!

survived replied:
Probably it's multi-party-eddsa who needs to use `Scalar<Secp256k1>` notation? 🙂 `FE` and `GE` come from older version of `curv` lib

survived replied:
The next possible step of multi-party-ecdsa evolution is to make it generic over choice of curve. Ie. points and scalars gonna look like `Point<E>` and `Scalar<E>`, and it will work with any curve, not only secp256k1

Hrezaei replied:
> Probably it's multi-party-eddsa who needs to use Scalar<Secp256k1> notation? 🙂

😊 Yes, exactly it is!

> The next possible step of multi-party-ecdsa evolution is to make it generic over choice of curve...

OK, that would be great! But just out of curiosity, will the ECDSA over a curve other than secp256k1 be called still "ECDSA"? How to distinguish them from each other?

survived replied:
Sure, ECDSA is not limited to secp256k1

survived replied:
> How to distinguish them from each other?

 At application layer. If you have signature on bitcoin transaction, it's secp256k1 ecdsa. If you got signature in TLS handshake, it's probably secp256r1

Hrezaei replied:
Thank you for your response! 🙏🏻
